QuEST/src/CPU/QuEST_cpu.c

```diff
@@ -3871,3 +3871,57 @@ void agnostic_setDiagonalOpElems(DiagonalOp op, long long int startInd, qreal* r
         }
     }
 }
+
+int hasEvenParity(long long x, int* in_qubitIndices, int qs){
+
+	size_t count = 0;
+	for (int i = 0; i < qs; ++i) {
+		int bitIdx = in_qubitIndices[i];
+		if (x & (1ULL << bitIdx))
+			count++;
+	}
+	return (count % 2) == 0;
+};
+
+void agnostic_initDiagonalOpFromPauliHamil(DiagonalOp op, PauliHamil hamil, qreal* offset){
```

Comment:
since processing more info than exists in a partitioned state, this function must be parallelised and hence moved into backend files